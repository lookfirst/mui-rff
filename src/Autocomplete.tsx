import React, { ChangeEvent, ReactNode } from 'react';

import { Field, FieldRenderProps, FieldProps } from 'react-final-form';

import TextField, { TextFieldProps as MuiTextFieldProps } from '@material-ui/core/TextField';
import {
	default as MuiAutocomplete,
	AutocompleteProps as MuiAutocompleteProps,
	RenderInputParams as MuiAutocompleteRenderInputParams,
} from '@material-ui/lab/Autocomplete';
import { showErrorOnChange } from './Util';

export type AutocompleteData = {
	[key: string]: any | null;
};

export interface AutocompleteProps extends Partial<Omit<MuiAutocompleteProps<any>, 'onChange'>> {
	name: string;
	label: ReactNode;
	helperText?: string;
	required?: boolean;
	multiple?: boolean;
	getOptionValue?: (option: any) => any;
	options: AutocompleteData[];
	fieldProps?: Partial<FieldProps<any, any>>;
	textFieldProps?: Partial<MuiTextFieldProps>;
}

export const Autocomplete = (props: AutocompleteProps) => {
	const { name, fieldProps, ...rest } = props;

	return (
		<Field
			name={name}
			render={(fieldRenderProps) => <AutocompleteWrapper {...fieldRenderProps} {...rest} />}
			{...fieldProps}
		/>
	);
};

interface AutocompleteWrapperProps extends FieldRenderProps<MuiTextFieldProps, HTMLElement> {
	label: ReactNode;
	required?: boolean;
	multiple?: boolean;
	textFieldProps?: Partial<MuiTextFieldProps>;
	getOptionValue?: (option: any) => any;
}

const AutocompleteWrapper = (props: AutocompleteWrapperProps) => {
	const {
		input: { name, onChange, value, ...restInput },
		meta,
		options,
		label,
		required,
		multiple,
		textFieldProps,
		getOptionValue,
		showError = showErrorOnChange,
		...rest
	} = props;

	function getValue(values: any) {
		if (!getOptionValue) {
			return values;
		}

		// ternary hell...
		return multiple ? (values ? values.map(getOptionValue) : null) : values ? getOptionValue(values) : null;
	}

	const { helperText, ...lessrest } = rest;
	const { variant, ...restTextFieldProps } = (textFieldProps as any) || {};

	// yuck...
	let defaultValue: any = null;

	if (!getOptionValue) {
		defaultValue = value;
	} else if (value !== null) {
		options.forEach((option: any) => {
			const optionValue = getOptionValue(option);
			if (multiple) {
				if (!defaultValue) defaultValue = [];
				(value as any).forEach((v: any) => {
					if (v === optionValue) {
						defaultValue.push(option);
					}
				});
			} else {
				if (value === optionValue) {
					defaultValue = option;
				}
			}
		});
	}

	const onChangeFunc = (_e: ChangeEvent<{}>, values: any | any[]) => onChange(getValue(values));

	const { error, submitError } = meta;
	const isError = showError({ meta });

	return (
		<MuiAutocomplete
			multiple={multiple}
			onChange={onChangeFunc}
			options={options}
			value={defaultValue}
			renderInput={(params: MuiAutocompleteRenderInputParams) => (
				<TextField
					label={label}
					required={required}
					fullWidth={true}
					helperText={isError ? error || submitError : helperText}
					error={isError}
					name={name}
					variant={variant}
					inputProps={{ required, ...restInput }}
					{...params}
					{...restTextFieldProps}
				/>
			)}
			{...lessrest}
		/>
	);
};
