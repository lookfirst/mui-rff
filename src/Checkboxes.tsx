import {
	Checkbox as MuiCheckbox,
	CheckboxProps as MuiCheckboxProps,
	FormControl,
	FormControlLabel,
	FormControlLabelProps,
	FormControlProps,
	FormGroup,
	FormGroupProps,
	FormHelperTextProps,
	FormLabel,
	FormLabelProps,
} from '@material-ui/core';
import React, { ReactNode } from 'react';

import { Field, FieldProps } from 'react-final-form';
import { ErrorMessage, showErrorOnChange, useFieldForErrors } from './Util';

export interface CheckboxData {
	label: ReactNode;
	value: unknown;
	disabled?: boolean;
}

export interface CheckboxesProps extends Partial<Omit<MuiCheckboxProps, 'onChange'>> {
	name: string;
	data: CheckboxData | CheckboxData[];
	label?: ReactNode;
	required?: boolean;
	helperText?: string;
	fieldProps?: Partial<FieldProps<any, any>>;
	formControlProps?: Partial<FormControlProps>;
	formGroupProps?: Partial<FormGroupProps>;
	formLabelProps?: Partial<FormLabelProps>;
	formControlLabelProps?: Partial<FormControlLabelProps>;
	formHelperTextProps?: Partial<FormHelperTextProps>;
	showError?: Function;
}

export function Checkboxes(props: CheckboxesProps) {
	const {
		required,
		label,
		data,
		name,
		helperText,
		fieldProps,
		formControlProps,
		formGroupProps,
		formLabelProps,
		formControlLabelProps,
		formHelperTextProps,
		showError = showErrorOnChange,
		...restCheckboxes
	} = props;

	const itemsData = !Array.isArray(data) ? [data] : data;
	const single = itemsData.length === 1;
	const field = useFieldForErrors(name);
	const isError = showError(field);

	return (
		<FormControl required={required} error={isError} {...formControlProps}>
			{label ? <FormLabel {...formLabelProps}>{label}</FormLabel> : <></>}
			<FormGroup {...formGroupProps}>
				{itemsData.map((item: CheckboxData, idx: number) => (
					<FormControlLabel
						key={idx}
						name={name}
						label={item.label}
						value={single ? undefined : item.value}
						disabled={item.disabled}
						control={
							<Field
								type="checkbox"
								name={name}
								render={({ input: { name, value, onChange, checked, ...restInput } }) => (
									<MuiCheckbox
										name={name}
										value={value}
										onChange={onChange}
										checked={checked}
										disabled={item.disabled}
										inputProps={{ required, ...restInput }}
										{...restCheckboxes}
									/>
								)}
								{...fieldProps}
							/>
						}
						{...formControlLabelProps}
					/>
				))}
			</FormGroup>
			<ErrorMessage
				showError={isError}
				meta={field.meta}
				formHelperTextProps={formHelperTextProps}
				helperText={helperText}
			/>
		</FormControl>
	);
}
