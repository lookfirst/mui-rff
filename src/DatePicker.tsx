import React from 'react';

import { DatePicker as MuiDatePicker, DatePickerProps as MuiDatePickerProps } from '@material-ui/pickers';

import { Field, FieldProps, FieldRenderProps } from 'react-final-form';

import { showErrorOnChange } from './Util';
import pickerProviderWrapper from './PickerProvider';

export interface DatePickerProps extends Partial<Omit<MuiDatePickerProps, 'onChange'>> {
	dateFunsUtils?: any;
	fieldProps?: Partial<FieldProps<any, any>>;
}

export function DatePicker(props: DatePickerProps) {
	const { name, fieldProps, ...rest } = props;

	return (
		<Field
			name={name as any}
			render={(fieldRenderProps) => <DatePickerWrapper {...fieldRenderProps} {...rest} />}
			{...fieldProps}
		/>
	);
}

interface DatePickerWrapperProps extends FieldRenderProps<MuiDatePickerProps, HTMLElement> {
	dateFunsUtils?: any;
}

function DatePickerWrapper(props: DatePickerWrapperProps) {
	const {
		input: { name, onChange, value, ...restInput },
		meta,
		dateFunsUtils,
		showError = showErrorOnChange,

		...rest
	} = props;

	const { error, submitError } = meta;
	const isError = showError({ meta });

	const { helperText, ...lessrest } = rest;

	return pickerProviderWrapper(
		dateFunsUtils,
		<MuiDatePicker
			fullWidth={true}
			autoOk={true}
			helperText={isError ? error || submitError : helperText}
			error={isError}
			onChange={onChange}
			name={name}
			value={(value as any) === '' ? null : value}
			{...lessrest}
			inputProps={restInput}
		/>,
	);
}
