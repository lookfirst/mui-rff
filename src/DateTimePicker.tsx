import React from 'react';

import {
	DateTimePicker as MuiDateTimePicker,
	DateTimePickerProps as MuiDateTimePickerProps,
} from '@material-ui/pickers';

import { Field, FieldProps, FieldRenderProps } from 'react-final-form';

import { showErrorOnChange } from './Util';
import pickerProviderWrapper from './PickerProvider';

export interface DateTimePickerProps extends Partial<Omit<MuiDateTimePickerProps, 'onChange'>> {
	dateFunsUtils?: any;
	fieldProps?: Partial<FieldProps<any, any>>;
}

export function DateTimePicker(props: DateTimePickerProps) {
	const { name, fieldProps, ...rest } = props;

	return (
		<Field
			name={name as any}
			render={fieldRenderProps => <DateTimePickerWrapper {...fieldRenderProps} {...rest} />}
			{...fieldProps}
		/>
	);
}

interface DateTimePickerWrapperProps extends FieldRenderProps<MuiDateTimePickerProps, HTMLElement> {
	dateFunsUtils?: any;
}

function DateTimePickerWrapper(props: DateTimePickerWrapperProps) {
	const {
		input: { name, onChange, value, ...restInput },
		meta,
		dateFunsUtils,
		showError = showErrorOnChange,
		...rest
	} = props;

	const { error, submitError } = meta;
	const isError = showError({ meta });

	const { helperText, ...lessrest } = rest;

	return pickerProviderWrapper(
		dateFunsUtils,
		<MuiDateTimePicker
			fullWidth={true}
			autoOk={true}
			helperText={isError ? error || submitError : helperText}
			error={isError}
			onChange={onChange}
			name={name}
			value={(value as any) === '' ? null : value}
			{...lessrest}
			inputProps={restInput}
		/>,
	);
}
