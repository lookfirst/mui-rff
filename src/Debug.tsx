import { FormSpy } from 'react-final-form';

export function Debug() {
	return (
		<FormSpy subscription={{ values: true }}>
			{({ values }) => <pre>{JSON.stringify(values, undefined, 2)}</pre>}
		</FormSpy>
	);
}
