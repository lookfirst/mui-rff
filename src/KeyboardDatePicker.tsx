import React from 'react';

import {
	KeyboardDatePicker as MuiKeyboardDatePicker,
	KeyboardDatePickerProps as MuiKeyboardDatePickerProps,
} from '@material-ui/pickers';

import { Field, FieldProps, FieldRenderProps } from 'react-final-form';

import { showErrorOnChange } from './Util';
import pickerProviderWrapper from './PickerProvider';

export interface KeyboardDatePickerProps extends Partial<Omit<MuiKeyboardDatePickerProps, 'onChange'>> {
	dateFunsUtils?: any;
	fieldProps?: Partial<FieldProps<any, any>>;
}

export function KeyboardDatePicker(props: KeyboardDatePickerProps) {
	const { name, fieldProps, ...rest } = props;

	return (
		<Field
			name={name as any}
			render={(fieldRenderProps) => <KeyboardDatePickerWrapper {...fieldRenderProps} {...rest} />}
			{...fieldProps}
		/>
	);
}

interface DatePickerWrapperProps extends FieldRenderProps<MuiKeyboardDatePickerProps, HTMLElement> {
	dateFunsUtils?: any;
}

function KeyboardDatePickerWrapper(props: DatePickerWrapperProps) {
	const {
		input: { name, onChange, value, ...restInput },
		meta,
		dateFunsUtils,
		...rest
	} = props;

	const { error, submitError } = meta;
	const isError = showError({ meta });

	const { helperText, ...lessrest } = rest;

	return pickerProviderWrapper(
		dateFunsUtils,
		<MuiKeyboardDatePicker
			disableToolbar
			fullWidth={true}
			autoOk={true}
			helperText={isError ? error || submitError : helperText}
			error={isError}
			onChange={onChange}
			name={name}
			value={(value as any) === '' ? null : value}
			inputProps={restInput}
			{...lessrest}
		/>,
	);
}
