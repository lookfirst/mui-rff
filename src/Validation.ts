import { ReactNode } from 'react';
import { Schema as YupSchema, ValidationError as YupValidationError } from 'yup';

// https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_get
function get(obj: any, path: string, defaultValue?: any) {
	const result = String.prototype.split
		.call(path, /[,[\].]+?/)
		.filter(Boolean)
		.reduce((res, key) => (res !== null && res !== undefined ? res[key] : res), obj);
	return result === undefined || result === obj ? defaultValue : result;
}

// https://stackoverflow.com/questions/54733539/javascript-implementation-of-lodash-set-method
function set(obj: any, path: any, value: any) {
	if (Object(obj) !== obj) return obj; // When obj is not an object
	// If not yet an array, get the keys from the string-path
	if (!Array.isArray(path)) path = path.toString().match(/[^.[\]]+/g) || [];
	path.slice(0, -1).reduce(
		(
			a: any,
			c: any,
			i: number, // Iterate all of them except the last one
		) =>
			Object(a[c]) === a[c] // Does the key exist and is its value an object?
				? // Yes: then follow that path
				  a[c]
				: // No: create the key. Is the next key a potential array-index?
				  (a[c] =
						Math.abs(path[i + 1]) >> 0 === +path[i + 1]
							? [] // Yes: assign a new array object
							: {}), // No: assign a new plain object
		obj,
	)[path[path.length - 1]] = value; // Finally assign the value to the last key
	return obj; // Return the top-level object to allow chaining
}

export type Translator = (errorObj: YupValidationError) => string | ReactNode;

export interface ValidationError {
	[key: string]: ValidationError | string;
}

function normalizeValidationError(err: YupValidationError, translator?: Translator): ValidationError {
	return err.inner.reduce((errors, innerError) => {
		let el: ReturnType<Translator>;

		const { path, message } = innerError;

		el = translator ? translator(innerError) : message;

		if (errors.hasOwnProperty(path)) {
			const prev = get(errors, path);
			prev.push(el);
			set(errors, path, prev);
		} else {
			set(errors, path, [el]);
		}
		return errors;
	}, {});
}

/**
 * Wraps the execution of a Yup schema to return an Promise<ValidationError>
 * where the key is the form field and the value is the error string.
 */
export function makeValidate<T>(validator: YupSchema<T>, translator?: Translator) {
	return async (values: T): Promise<ValidationError> => {
		try {
			await validator.validate(values, { abortEarly: false });
			return {};
		} catch (err) {
			return normalizeValidationError(err, translator);
		}
	};
}

/**
 * Wraps the sync execution of a Yup schema to return an ValidationError
 * where the key is the form field and the value is the error string.
 */
export function makeValidateSync<T>(validator: YupSchema<T>, translator?: Translator) {
	return (values: T): ValidationError => {
		try {
			validator.validateSync(values, { abortEarly: false });
			return {};
		} catch (err) {
			return normalizeValidationError(err, translator);
		}
	};
}

/**
 * Uses the private _exclusive field in the schema to get whether or not
 * the field is marked as required or not.
 */
export function makeRequired<T>(schema: YupSchema<T>) {
	const fields = (schema as any).fields;
	return Object.keys(fields).reduce((accu, field) => {
		if (fields[field].fields) {
			accu[field] = makeRequired(fields[field]);
		} else {
			accu[field] = !!fields[field]._exclusive.required;
		}
		return accu;
	}, {} as any);
}
