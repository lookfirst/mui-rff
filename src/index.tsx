export { Autocomplete, AutocompleteData, AutocompleteProps } from './Autocomplete';
export { Checkboxes, CheckboxData, CheckboxesProps } from './Checkboxes';
export { Switches, SwitchData, SwitchesProps as SwitchProps } from './Switches';
export { Radios, RadioData, RadiosProps } from './Radios';
export { Select, SelectData, SelectProps } from './Select';
export { KeyboardDatePicker, KeyboardDatePickerProps } from './KeyboardDatePicker';
export { KeyboardDateTimePicker, KeyboardDateTimePickerProps } from './KeyboardDateTimePicker';
export { DatePicker, DatePickerProps } from './DatePicker';
export { DateTimePicker, DateTimePickerProps } from './DateTimePicker';
export { KeyboardTimePicker, KeyboardTimePickerProps } from './KeyboardTimePicker';
export { TimePicker, TimePickerProps } from './TimePicker';
export { TextField, TextFieldProps } from './TextField';
export { makeValidate, makeValidateSync, makeRequired } from './Validation';
export { Debug } from './Debug';
export {
	ErrorMessageProps,
	ErrorMessage,
	ShowErrorProps,
	showErrorOnBlur,
	showErrorOnChange,
	useFieldForErrors,
} from './Util';
