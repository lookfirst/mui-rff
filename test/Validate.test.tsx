import React from 'react';

import { Form } from 'react-final-form';
import * as Yup from 'yup';

import { makeValidate, TextField } from '../src';
import { customRender, fireEvent, getNodeText } from './TestUtils';
import { Translator } from '../src/Validation';

Yup.setLocale({
	mixed: {
		required: ({ path }) => ({
			key: 'field_required',
			field: path,
		}),
	},
	string: {
		min: ({ path }) => ({
			key: 'field_too_short',
			field: path,
		}),
		email: ({ path }) => ({
			key: 'field_not_email',
			field: path,
		}),
	},
});

const myTranslatorFunction: Translator = ({ message }: Yup.ValidationError) => {
	const error = message as any;
	// use some kind of translation library to actually translate the objects to strings (like i18next)
	return `${error.key}: ${error.field}`;
};

const myExtendedTranslatorFunction: Translator = ({ message }: Yup.ValidationError) => {
	const error = message as any;
	// use some kind of translation library to actually translate the objects to strings (like i18next)
	return (
		<span data-testid="error_field" key={error.key}>
			{error.key}: {error.field}
		</span>
	);
};

interface ComponentProps {
	initialValues: FormData;
	validator?: any;
	onSubmit?: any;
}

interface FormData {
	hello: string;
}

describe('Validate', () => {
	describe('multiple validation errors', () => {
		const initialValues: FormData = {
			hello: '',
		};

		function TextFieldComponent({ initialValues, validator }: ComponentProps) {
			const onSubmit = (values: FormData) => {
				console.log(values);
			};

			const validate = async (values: FormData) => {
				if (validator) {
					return validator(values);
				}
			};

			return (
				<Form
					onSubmit={onSubmit}
					initialValues={initialValues}
					validate={validate}
					render={({ handleSubmit }) => (
						<form onSubmit={handleSubmit} noValidate>
							<TextField
								label="Test"
								name="hello"
								required={true}
								inputProps={{ 'data-testid': 'textfield' }}
							/>
						</form>
					)}
				/>
			);
		}

		it('with YUP localisation mingles objects when no translator', async () => {
			const validateSchema = makeValidate(
				Yup.object().shape({
					hello: Yup.string().required().min(10).email(),
				}),
			);

			const dataFaulty = {
				hello: '',
			};

			const errors = await validateSchema(dataFaulty);

			expect(errors).toStrictEqual({
				hello: [
					{ field: 'hello', key: 'field_required' },
					{ field: 'hello', key: 'field_too_short' },
				],
			});
		});

		it('with YUP localisation doesnt mingle objects with a translator', async () => {
			const validateSchema = makeValidate(
				Yup.object().shape({
					hello: Yup.string().required().min(10).email(),
				}),
				myTranslatorFunction,
			);

			const dataFaulty = {
				hello: '',
			};

			const errors = await validateSchema(dataFaulty);

			expect(errors).toEqual({ hello: ['field_required: hello', 'field_too_short: hello'] });
		});

		it('can render multiple errors', async () => {
			const message = 'field_too_short: hellofield_not_email: hello';

			const validateSchema = makeValidate(
				Yup.object().shape({
					hello: Yup.string().required().min(10).email(),
				}),
				myTranslatorFunction,
			);

			const { findByText, container } = customRender(
				<TextFieldComponent validator={validateSchema} initialValues={initialValues} />,
			);
			const input = container.querySelector('input') as HTMLInputElement;

			// ensure the validation is made and errors are rendered
			fireEvent.focus(input);
			fireEvent.change(input, {
				target: { value: 'no email' },
			});
			fireEvent.blur(input);

			// find element with errors
			const error = await findByText(/field_not_email/); // validation is async, so we have to await
			expect(error.innerHTML).toContain(message);

			expect(container).toMatchSnapshot();
		});

		it('can render multiple errors in separate elements', async () => {
			const validateSchema = makeValidate(
				Yup.object().shape({
					hello: Yup.string().required().min(10).email(),
				}),
				myExtendedTranslatorFunction,
			);

			const { findAllByTestId, container } = customRender(
				<TextFieldComponent validator={validateSchema} initialValues={initialValues} />,
			);
			const input = container.querySelector('input') as HTMLInputElement;

			// ensure the validation is made and errors are rendered
			fireEvent.focus(input);
			fireEvent.change(input, {
				target: { value: 'no email' },
			});
			fireEvent.blur(input);

			// find error fields
			const errors = await findAllByTestId('error_field'); // validation is async, so we have to await
			expect(errors).toHaveLength(2);
			expect(getNodeText(errors[0])).toContain('field_too_short: hello');
			expect(getNodeText(errors[1])).toContain('field_not_email: hello');

			expect(container).toMatchSnapshot();
		});
	});
});
