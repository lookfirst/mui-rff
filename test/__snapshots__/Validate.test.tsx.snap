// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate multiple validation errors can render multiple errors 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="MuiFormControl-root-2 MuiTextField-root-1 MuiFormControl-fullWidth-5"
    >
      <label
        class="MuiFormLabel-root-18 MuiInputLabel-root-6 MuiInputLabel-formControl-12 MuiInputLabel-animated-15 MuiInputLabel-shrink-14 MuiFormLabel-error-22 MuiInputLabel-error-9 MuiFormLabel-filled-23 MuiFormLabel-required-24 MuiInputLabel-required-10"
        data-shrink="true"
      >
        Test
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk-25 MuiInputLabel-asterisk-11 MuiFormLabel-error-22 MuiInputLabel-error-9"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root-40 MuiInput-root-26 MuiInput-underline-31 MuiInputBase-error-46 MuiInput-error-32 MuiInputBase-fullWidth-50 MuiInput-fullWidth-35 MuiInputBase-formControl-41 MuiInput-formControl-27"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input-51 MuiInput-input-36"
          data-testid="textfield"
          name="hello"
          required=""
          type="text"
          value="no email"
        />
      </div>
      <p
        class="MuiFormHelperText-root-58 MuiFormHelperText-error-59 MuiFormHelperText-filled-64 MuiFormHelperText-required-65"
      >
        field_too_short: hello
        field_not_email: hello
      </p>
    </div>
  </form>
</div>
`;

exports[`Validate multiple validation errors can render multiple errors in separate elements 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="MuiFormControl-root-2 MuiTextField-root-1 MuiFormControl-fullWidth-5"
    >
      <label
        class="MuiFormLabel-root-18 MuiInputLabel-root-6 MuiInputLabel-formControl-12 MuiInputLabel-animated-15 MuiInputLabel-shrink-14 MuiFormLabel-error-22 MuiInputLabel-error-9 MuiFormLabel-filled-23 MuiFormLabel-required-24 MuiInputLabel-required-10"
        data-shrink="true"
      >
        Test
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk-25 MuiInputLabel-asterisk-11 MuiFormLabel-error-22 MuiInputLabel-error-9"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root-40 MuiInput-root-26 MuiInput-underline-31 MuiInputBase-error-46 MuiInput-error-32 MuiInputBase-fullWidth-50 MuiInput-fullWidth-35 MuiInputBase-formControl-41 MuiInput-formControl-27"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input-51 MuiInput-input-36"
          data-testid="textfield"
          name="hello"
          required=""
          type="text"
          value="no email"
        />
      </div>
      <p
        class="MuiFormHelperText-root-58 MuiFormHelperText-error-59 MuiFormHelperText-filled-64 MuiFormHelperText-required-65"
      >
        <span
          data-testid="error_field"
        >
          field_too_short
          : 
          hello
        </span>
        <span
          data-testid="error_field"
        >
          field_not_email
          : 
          hello
        </span>
      </p>
    </div>
  </form>
</div>
`;
